src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt

```diff
@@ -582,46 +601,51 @@ abstract class ManagedStrategy(
+    open fun onThreadStart(threadId: ThreadId) {
+        threadScheduler.awaitTurn(threadId)
+        threadScheduler.startThread(threadId)
     }
 
     /**
-     * This method is executed as the last thread action if no exception has been thrown.
-     * @param iThread the number of the executed thread according to the [scenario][ExecutionScenario].
+     * This method is executed as the last thread action.
+     *
+     * @param threadId the thread id of the finished thread.
      */
-    open fun onThreadFinish(iThread: Int) {
-        threadScheduler.awaitTurn(iThread)
-        threadScheduler.finishThread(iThread)
-        loopDetector.onThreadFinish(iThread)
+    open fun onThreadFinish(threadId: ThreadId) {
+        threadScheduler.awaitTurn(threadId)
+        threadScheduler.finishThread(threadId)
+        loopDetector.onThreadFinish(threadId)
         traceCollector?.onThreadFinish()
-        unblockJoiningThreads(iThread)
-        val nextThread = chooseThreadSwitch(iThread, true)
+        unblockJoiningThreads(threadId)
+        val nextThread = chooseThreadSwitch(threadId, true)
         setCurrentThread(nextThread)
     }
 
     /**
-     * This method is executed if an illegal exception has been thrown (see [exceptionCanBeValidExecutionResult]).
-     * @param iThread the number of the executed thread according to the [scenario][ExecutionScenario].
-     * @param exception the exception that was thrown
+     * This method is executed if an internal exception has been thrown (see [isInternalException]).
+     *
+     * @param threadId the thread id of the thread where exception was thrown.
+     * @param exception the exception that was thrown.
+     * @return true if the exception should be suppressed, false otherwise.
      */
-    open fun onThreadFailure(iThread: Int, exception: Throwable) {
-        // This method is called only if exception can't be treated as a normal operation result,
+    open fun onInternalException(threadId: Int, exception: Throwable): Boolean {
+        // This method is called only if the exception cannot be treated as a normal result,
         // so we exit testing code to avoid trace collection resume or some bizarre bugs
         leaveTestingCode()
-        // skip abort exception
-        if (exception !== ThreadAbortedError) {
-            // Despite the fact that the corresponding failure will be detected by the runner,
-            // the managed strategy can construct a trace to reproduce this failure.
-            // Let's then store the corresponding failing result and construct the trace.
-            suddenInvocationResult = UnexpectedExceptionInvocationResult(exception, runner.collectExecutionResults())
-            threadScheduler.abortAllThreads()
-        }
-        // notify the scheduler that the thread is going to be finished
-        threadScheduler.finishThread(iThread)
+        // suppress `ThreadAbortedError`
+        if (exception is ThreadAbortedError) return true
```

Comment:
Let's rename `ThreadAbortedError` to `LincheckAnalysisAbortedError` or smth similar, to highlight how we use it.